src/handler.rs

```diff
@@ -0,0 +1,234 @@
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.",
+    )
+  }
+
+  fn format_text(
+    &mut self,
+    file_path: &Path,
+    _file_text: &str,
+    config: &Configuration,
+    mut _format_with_host: impl FnMut(&Path, String, &ConfigKeyMap) -> Result<String>,
+  ) -> Result<String> {
+    format_text(file_path, _file_text, config, _format_with_host)
+  }
+}
+
+#[tokio::main]
+pub async fn format_text(
+  file_path: &Path,
+  file_text: &str,
+  config: &Configuration,
+  mut _format_with_host: impl FnMut(&Path, String, &ConfigKeyMap) -> Result<String>,
+) -> Result<String> {
+  // format here
+  let args = maybe_substitute_variables(file_path, file_text, &config);
+
+  let mut child = Command::new(&config.executable)
+    .kill_on_drop(true)
+    .current_dir(&config.exe_dir_path)
+    .stdout(Stdio::piped())
+    .stdin(if config.stdin {
+      Stdio::piped()
+    } else {
+      Stdio::null()
+    })
+    .stderr(Stdio::piped())
+    .args(args)
+    .spawn()
+    .map_err(|e| anyhow!("Cannot start formatter process: {}", e))?;
+
+  // capturing stdout
+  let (out_tx, out_rx) = oneshot::channel();
+  if let Some(stdout) = child.stdout.take() {
+    let eol = resolve_new_line_kind(file_text, config.new_line_kind);
+    tokio::spawn(read_stream_lines(stdout, eol, out_tx));
+  } else {
+    return Err(anyhow!("Formatter did not have a handle for stdout"));
+  }
+
+  // capturing stderr
+  let (err_tx, err_rx) = oneshot::channel();
+  if let Some(stderr) = child.stderr.take() {
+    let system_eol = resolve_new_line_kind(file_text, NewLineKind::System);
+    tokio::spawn(read_stream_lines(stderr, system_eol, err_tx));
+  }
+
+  // write file text into child's stdin
+  if config.stdin {
+    child
+      .stdin
+      .take()
+      .ok_or_else(|| anyhow!("Cannot open formatter stdin"))?
+      .write_all(file_text.as_bytes())
+      .await
+      .map_err(|_| anyhow!("Cannot write into formatter stdin"))
+      .unwrap();
+  }
+
+  // Ensure the child process is spawned in the runtime so it can
+  // make progress on its own while we await for any output.
+  let child_completed = tokio::spawn(async move {
+    match child.wait().await {
+      Ok(status) => status,
+      Err(e) => panic!("Error while waiting for formatter to complete: {}", e),
+    }
+  });
+
+  match timeout(Duration::from_secs(config.timeout as u64), out_rx).await {
+    Ok(result) => handle_child_exit_status(result?, err_rx, child_completed.await?).await,
+    Err(e) => Err(timeout_err(config, e)),
+  }
```

Comment:
I have a prepared PR in the private repo where I adding support of so called "in-place" formatters. These types of formatters do not return anything via `stdout`.
We do not need this feature right now and generally cannot come to agreement is this feature needed or not. If you think it is needed I can open another PR.

Reply:
I think I'd rather just support formatters that can format via stdin unless there's a very popular one that doesn't support this. Do you know any formatters that can't?